### PR TITLE
Support custom platforms

### PR DESCRIPTION
**Pull request by ajeddeloh**
Add the ability to bring your own metadata service. Add a "custom" platform type and document how to use it by overwriting parts of the coreos-metadata service

### CODE REVIEW

File: doc/dynamic-data.md [hunk 1]
```diff
@@ -24,6 +24,49 @@ This is the information available in each provider.
 | OpenStack-Metadata | ✓          | ✓              | ✓             |                |               |
 | Vagrant-Virtualbox | ✓          | ✓              |               |                |               |
 
+## Custom metadata providers
+
+As of ct 7.0(FIXME: what do we put here), ct also supports custom metadata providers. To use the `custom` platform, overwrite the coreos-metadata service unit to execute your own custom metadata fetcher. The custom metadata fetcher much write an enviroment file with all of the dynamic fields used to `/run/metadata/coreos`.
```
Comment (dgonyeo) [inline doc/dynamic-data.md]: imo we don't need to say what version this exists in, just release a new version with this quick-ish and if people ask we can show them the NEWS file.

Comment (bgilbert) [reply doc/dynamic-data.md]: s/much/must/
s/enviroment/environment/
`dynamic fields used to /run/metadata/coreos`?

File: doc/dynamic-data.md [hunk 2]
```diff
@@ -24,6 +24,49 @@ This is the information available in each provider.
...
+PUBLIC_IPV4=<The instance's public ipv4 address>
+```
+
+The following Container Linux Config downloads the metadata fetching script, replaces the ExecStart line in `coreos-metadata` service to use the script instead, and configures etcd using the metadata provided.
+```yaml
+storage:
+  files:
+    - filesystem: "root"
+      path: "/opt/get-metadata.sh"
+      mode: 0766
```
Comment (bgilbert) [inline doc/dynamic-data.md]: World-writable?

**Comment by dgonyeo** [comment]
The shell script could even be sourced from the local disk at conversion time with the recent local files feature that was added

File: doc/dynamic-data.md [hunk 3]
```diff
@@ -24,6 +24,49 @@ This is the information available in each provider.
...
+          url: "https://example.com/metadata-script.sh"
+
+systemd:
+  units:
+    - name: "coreos-metadata.service"
+      dropins:
+       - name: "use-script.conf"
+         contents: |
+           [Service]
+           ExecStart=
```
Comment (bgilbert) [inline doc/dynamic-data.md]: Needs a comment.

**Comment by bgilbert** [comment]
As discussed OOB, we're going to leave out the sugar for now to avoid creating a compatibility constraint, but may add it later once we have more experience with the feature.

**Comment by ajeddeloh** [comment]
Addressed various typos and fat-fingerings of file modes. Added some comments and made the language a little clearer.

File: config/templating/templating.go [hunk 4]
```diff
@@ -76,6 +76,13 @@ var platformTemplatingMap = map[string]map[string]string{
 	platform.CloudStackConfigDrive: {
 		fieldHostname: "CLOUDSTACK_LOCAL_HOSTNAME",
 	},
+	platform.Custom: {
+		fieldHostname:  "HOSTNAME",
```
Comment (bgilbert) [inline config/templating/templating.go]: Should the variable names be namespaced like the others?

Comment (ajeddeloh) [reply config/templating/templating.go]: I was unsure. From an implementation standpoint, it makes a lot of sense. From a user I think it makes more sense to use what they are using in their templated config so they don't have to mental map them. It's just a little bit more they need to understand (although it is documented below). Either way I don't feel strongly.

**Comment by ajeddeloh** [comment]
Updated to namedspace everything with `COREOS_` and updated the docs to match.

File: doc/dynamic-data.md [hunk 5]
```diff
@@ -24,6 +24,51 @@ This is the information available in each provider.
 | OpenStack-Metadata | ✓          | ✓              | ✓             |                |               |
 | Vagrant-Virtualbox | ✓          | ✓              |               |                |               |
 
+## Custom metadata providers
+
+`ct` also supports custom metadata providers. To use the `custom` platform, modify the coreos-metadata service unit to execute your own custom metadata fetcher. The custom metadata fetcher must write an environment file to `/run/metadata/coreos` which defines all the above environment variables used in the Container Linux Config. The environment variables are the same as in the Container Linux Config, but prefixed with `COREOS_`.
```
Comment (bgilbert) [inline doc/dynamic-data.md]: `COREOS_CUSTOM_` now.

What happens if only some variables are defined?  There might not be all four of public/private IPv4/IPv6.

Comment (ajeddeloh) [reply doc/dynamic-data.md]: Oops, will fix that.

`ct` has no way of knowing what the custom metadata service will be able to provide, so it always allows them. I just tested and systemd will resolve an undefined environment variable to an empty string.

Comment (ajeddeloh) [reply doc/dynamic-data.md]: My intent with that sentence was to say that everything the user uses in their CLC must written out by the custom metadata service. What about:

"The custom metadata fetcher must write an environment file to `/run/metadata/coreos` defining every `COREOS_CUSTOM_*` environment variable used in the specified Container Linux Config"

Comment (ajeddeloh) [reply doc/dynamic-data.md]: Changed to that.

File: doc/dynamic-data.md [hunk 6]
```diff
@@ -24,6 +24,51 @@ This is the information available in each provider.
 | OpenStack-Metadata | ✓          | ✓              | ✓             |                |               |
 | Vagrant-Virtualbox | ✓          | ✓              |               |                |               |
 
+## Custom metadata providers
+
+`ct` also supports custom metadata providers. To use the `custom` platform, modify the coreos-metadata service unit to execute your own custom metadata fetcher. The custom metadata fetcher must write and environment file `/run/metadata/coreos` defining a `COREOS_CUSTOM_*` environment variable for every piece of dynamic data used in the specified Container Linux Config. The environment variables are the same as in the Container Linux Config, but prefixed with `COREOS_CUSTOM_`.
```
Comment (bgilbert) [inline doc/dynamic-data.md]: `write an`

File: doc/dynamic-data.md [hunk 7]
```diff
@@ -24,6 +24,51 @@ This is the information available in each provider.
...
+
+Example:
```
Comment (bgilbert) [inline doc/dynamic-data.md]:
```
### Example
```

File: doc/dynamic-data.md [hunk 8]
```diff
@@ -24,6 +24,51 @@ This is the information available in each provider.
...
+
+This example assumes `https://example.com/metadata-script.sh` is a script which communicates with a metadata service then writes the following file to `/run/metadata/coreos`:
```
Comment (bgilbert) [inline doc/dynamic-data.md]: s/This example assumes/Assume/
s/then/and then/

File: doc/dynamic-data.md [hunk 9]
```diff
@@ -24,6 +24,51 @@ This is the information available in each provider.
...
+```
+COREOS_CUSTOM_HOSTNAME=foobar
+COREOS_CUSTOM_PRIVATE_IPV4=<The instance's private ipv4 address>
+COREOS_CUSTOM_PUBLIC_IPV4=<The instance's public ipv4 address>
+```
+
+The following Container Linux Config downloads the metadata fetching script, replaces the ExecStart line in `coreos-metadata` service to use the script instead, and configures etcd using the metadata provided. Use the `--platform=custom` flag when transpiling.
+```yaml
```
Comment (bgilbert) [inline doc/dynamic-data.md]: Should this be `yaml container-linux-config:custom`?

**Comment by ajeddeloh** [comment]
Updated the grammar. Left off the `:custom` for the config type until the next CT release happens (since the docs sync doesn't use master IIRC)

**Comment by dghubble** [comment]
If I understand this correctly, this is targeting a user who:

* writes a custom metadata fetching script
* Adds that script on the filesystem
* Adds a coreos-metadata.service unit that uses that script

all so that the user can write `http://{PRIVATE_IPV4}:2379` instead of `http://{COREOS_CUSTOM_PRIVATE_IPV4}:2379` in the CLC. Was this a problem folks were having?

**Comment by ajeddeloh** [comment]
Yeah, there was a coreos-user thread about it. CT doesn't allow any `{ }` templating without specifying a platform, regardless of what the value inside the `{}` is. This also makes it so the same CLC can target baremetal and non-baremetal platforms.

**Pull request merged**